File: 2015/12/solution.js
const solve = (node, excludeObjectsWithValue) => {
  let nodeValue = 0
  if (typeof node === 'number') {
    return node
  } else if (typeof node === 'object') {
    if (Array.isArray(node)) {
      node.forEach((value) => nodeValue += solve(value, excludeObjectsWithValue))
    } else {
      const values = Object.values(node)
      for (let i = 0; i < values.length; i++) {
        if (excludeObjectsWithValue && values[i] === excludeObjectsWithValue) {
          return 0
        }
        nodeValue += solve(values[i], excludeObjectsWithValue)
      }
    }
  }
  return nodeValue
}

const input = require('fs').readFileSync('input.txt', 'utf8')
const json = JSON.parse(input)

console.log(solve(json))
console.log(solve(json, 'red'))

File: 2015/12/solution.rb
require 'json'

def solve(node, exclude_value = nil)
  node_value = 0
  return node if node.is_a? Numeric
  if node.is_a?(Array)
    node.each do |value|
      node_value += solve(value, exclude_value)
    end
  elsif node.respond_to? :values
    node.values.each do |value|
      return 0 if exclude_value && value == exclude_value
      node_value += solve(value, exclude_value)
    end
  end
  node_value
end

input = File.read('input.txt')
json = JSON.parse(input)

puts 'Part 1', solve(json)
puts 'Part 2', solve(json, 'red')
